examples_tests/__init__.py

```diff
@@ -0,0 +1,161 @@
+        Path of the most-recently-modified private SSH key
+    :raises LookupError:
+        If no such key was found.
+
+    This function tries to mimic the logic found in ``ubuntu-device-flash``.
+    It will look for the most recently modified private key in the users' SSH
+    configuration directory.
+    """
+    candidates = []
+    ssh_dir = os.path.expanduser('~/.ssh/')
+    for filename in os.listdir(ssh_dir):
+        # Skip public keys, we want the private key
+        if filename.endswith('.pub'):
+            continue
+        ssh_key = os.path.join(ssh_dir, filename)
+        # Skip non-files
+        if not os.path.isfile(ssh_key):
+            continue
+        # Ensure that it is a real ssh key
+        with open(ssh_key, 'rb') as stream:
+            if stream.readline() != b'-----BEGIN RSA PRIVATE KEY-----\n':
+                continue
+        candidates.append(ssh_key)
+    # Sort the keys by modification time, pick the most recent key
+    candidates.sort(key=lambda f: os.stat(f).st_mtime, reverse=True)
+    logger.debug('Available ssh public keys: %r', candidates)
+    if not candidates:
+        raise LookupError('Unable to find any private ssh key')
+    return candidates[0]
+
+
+class ExampleTestCase(testtools.TestCase):
+
+    example_dir = None
+
+    def setUp(self):
+        filter_ = config.get('filter', None)
+        if filter_:
+            if not re.match(filter_, self.example_dir):
+                self.skipTest(
+                    '{} does not match the filter {}'.format(
+                        self.example_dir, filter_))
+
+        super().setUp()
+        # To measure coverage, a wrapper for the snapcraft binary might be set
+        # in the environment variable.
+        snapcraft_bin = os.getenv('SNAPCRAFT', 'snapcraft')
+        self.snapcraft_command = os.path.join(
+            os.getcwd(), 'bin', snapcraft_bin)
+
+        self.useFixture(
+            fixtures.EnvironmentVariable('SNAPCRAFT_SETUP_PROXIES', '1'))
+
+        self.snappy_testbed = None
+        if not config.get('skip-install', False):
+            ip = config.get('ip', None)
+            if not ip:
+                self.snappy_testbed = self._set_up_qemu_testbed()
+            else:
+                port = config.get('port', None) or '22'
+                self.snappy_testbed = testbed.SshTestbed(
+                    ip, port, 'ubuntu')
+            self.snappy_testbed.wait()
+
+    def _set_up_qemu_testbed(self):
+        private_key = _get_latest_ssh_private_key()
+        snappy_image = config.get('snappy_image', None)
+        if not snappy_image:
+            temp_dir = tempfile.mkdtemp()
+            snappy_image = testbed.create_snappy_image(temp_dir)
+            # Store the image path in the config so it's only create once
+            # per execution.
+            config['snappy_image'] = snappy_image
+            # Delete the image when the execution exits.
+            atexit.register(shutil.rmtree, temp_dir)
```

Comment:
The idea is to create the image only once per execution, because it's what takes more time.
I tried using setupModule, but with this refactor it will mean once per example, which is also a lot of time. So the remaining options were to do it on __main__.py, or do a lazy creation and delete when the main thread exits. I liked a little more this last one, but I'm ok to change it if you have a better option in mind.